src/device_trezor/device_trezor_base.cpp

```diff
@@ -441,48 +453,89 @@ namespace trezor {
         pin = m_pin;
       }
 
-      // TODO: remove PIN from memory
+      std::string pin_field;
       messages::common::PinMatrixAck m;
       if (pin) {
-        m.set_pin(pin.get().data(), pin.get().size());
+        pin_field.assign(pin->data(), pin->size());
+        m.set_allocated_pin(&pin_field);
       }
+
+      const auto data_cleaner = epee::misc_utils::create_scope_leave_handler([&]() {
+        m.release_pin();
+        if (!pin_field.empty()){
+          memwipe(&pin_field[0], pin_field.size());
+        }
+      });
+
       resp = call_raw(&m);
     }
 
     void device_trezor_base::on_passphrase_request(GenericMessage & resp, const messages::common::PassphraseRequest * msg)
     {
       CHECK_AND_ASSERT_THROW_MES(msg, "Empty message");
-      MDEBUG("on_passhprase_request, on device: " << msg->on_device());
-      boost::optional<epee::wipeable_string> passphrase;
-      TREZOR_CALLBACK_GET(passphrase, on_passphrase_request, msg->on_device());
+      MDEBUG("on_passhprase_request");
 
-      if (!passphrase && m_passphrase){
-        passphrase = m_passphrase;
+      // Backward compatibility, migration clause.
+      if (msg->has__on_device() && msg->_on_device()){
+        messages::common::PassphraseAck m;
+        resp = call_raw(&m);
+        return;
       }
 
-      m_passphrase = boost::none;
+      bool on_device = true;
+      if (msg->has__on_device() && !msg->_on_device()){
+        on_device = false;  // do not enter on device, old devices.
+      }
 
-      messages::common::PassphraseAck m;
-      if (!msg->on_device() && passphrase){
-        // TODO: remove passphrase from memory
-        m.set_passphrase(passphrase.get().data(), passphrase.get().size());
+      if (on_device && m_features && m_features->capabilities_size() > 0){
+        on_device = false;
+        for (auto it = m_features->capabilities().begin(); it != m_features->capabilities().end(); it++) {
+          if (*it == messages::management::Features::Capability_PassphraseEntry){
+            on_device = true;
+          }
+        }
       }
 
-      if (!m_device_state.empty()){
-        m.set_allocated_state(&m_device_state);
+      boost::optional<epee::wipeable_string> passphrase;
+      TREZOR_CALLBACK_GET(passphrase, on_passphrase_request, on_device);
```

Comment:
What happens if the device doesn't support PassphraseEntry (i.e. invoking this with `on_device == false`), but the callback returns `on_device = true`?

Reply:
Currently, Monero is supported only by Trezor versions that support passphrase entry, i.e., model T. If callback returns `on_device = true` for device not supporting passprase entry, it would lead to error.

It is the same as we would throw exception in the code.